dio/lib/src/multipart_file.dart

```diff
@@ -12,21 +12,46 @@ import 'multipart_file/io_multipart_file.dart'
 /// MultipartFile is based on stream, and a stream can be read only once,
 /// so the same MultipartFile can't be read multiple times.
 class MultipartFile {
+  /// The size of the file in bytes. This must be known in advance, even if this
+  /// file is created from a [ByteStream].
+  final int length;
+
+  /// The basename of the file. May be null.
+  final String? filename;
+
+  /// The additional headers the file has. May be null.
+  final Map<String, List<String>>? headers;
+
+  /// The content-type of the file. Defaults to `application/octet-stream`.
+  final MediaType? contentType;
+
+  /// The stream that will emit the file's contents.
+  @Deprecated('Use [data] instead.')
+  final Stream<List<int>> _stream;
+
+  // The stream builder that will emit the file's contents for every call.
+  final Stream<List<int>> Function() _data;
+
+  /// Whether [finalize] has been called.
+  bool get isFinalized => _isFinalized;
+  bool _isFinalized = false;
+
   /// Creates a new [MultipartFile] from a chunked [Stream] of bytes. The length
   /// of the file in bytes must be known in advance. If it's not, read the data
   /// from the stream and use [MultipartFile.fromBytes] instead.
   ///
   /// [contentType] currently defaults to `application/octet-stream`, but in the
   /// future may be inferred from [filename].
   MultipartFile(
-    Stream<List<int>> stream,
+    Stream<List<int>> Function() data,
```

Comment:
Okay, will do. If it's a new factory, it can't be called by other factories, right? Should I remove the other factories and make `data`a named parameter, so that the old constructor can still work? I'm sorry I'm a bit confused with this

Reply:
Also, I believe what makes this solution work is that we're storing the stream generating method in `_data`, not the stream itself

Reply:
Your solution looks great, thanks for the help!